migrate_cpp/cpp_refactoring/rewriter.cpp

```diff
@@ -0,0 +1,286 @@
+      // write any.
+      return true;
+  }
+  Write(type_loc, OutputSegment::Text(content));
+  return true;
+}
+
+auto RewriteBuilder::VisitCXXBoolLiteralExpr(clang::CXXBoolLiteralExpr* expr)
+    -> bool {
+  Write(expr, OutputSegment::Text(expr->getValue() ? "true" : "false"));
+  return true;
+}
+
+auto RewriteBuilder::VisitDeclRefExpr(clang::DeclRefExpr* expr) -> bool {
+  Write(expr, OutputSegment::Text(TextForTokenAt(expr->getBeginLoc())));
+  return true;
+}
+
+auto RewriteBuilder::VisitDeclStmt(clang::DeclStmt* stmt) -> bool {
+  std::vector<OutputSegment> segments;
+  for (clang::Decl* decl : stmt->decls()) {
+    segments.push_back(OutputSegment::Node(decl));
+    segments.push_back(OutputSegment::Text(";\n"));
+  }
+  Write(stmt, std::move(segments));
+  return true;
+}
+
+auto RewriteBuilder::VisitIntegerLiteral(clang::IntegerLiteral* expr) -> bool {
+  // TODO: Replace suffixes.
+  std::string text(TextForTokenAt(expr->getBeginLoc()));
+  for (char& c : text) {
+    if (c == '\'') {
+      c = '_';
+    }
+  }
+  Write(expr, {OutputSegment::Text(std::move(text))});
+  return true;
+}
+
+auto RewriteBuilder::VisitPointerTypeLoc(clang::PointerTypeLoc type_loc)
+    -> bool {
+  Write(type_loc, {OutputSegment::TypeLoc(type_loc.getPointeeLoc()),
+                   OutputSegment::Text("*")});
+  return true;
+}
+
+auto RewriteBuilder::VisitTranslationUnitDecl(clang::TranslationUnitDecl* decl)
+    -> bool {
+  std::vector<OutputSegment> segments;
+
+  // Clang starts each translation unit with some initial `TypeDefDecl`s that
+  // are not part of the written text. We want to skip past these initial
+  // declarations, which we do by ignoring any node of type `TypeDefDecl` which
+  // has an invalid source location.
+  auto iter = decl->decls_begin();
+  while (iter != decl->decls_end() && llvm::isa<clang::TypedefDecl>(*iter) &&
+         (*iter)->getLocation().isInvalid()) {
+    ++iter;
+  }
+
+  for (; iter != decl->decls_end(); ++iter) {
+    clang::Decl* d = *iter;
+    segments.push_back(OutputSegment::Node(d));
+    segments.push_back(OutputSegment::Text(";\n"));
+  }
+
+  Write(decl, std::move(segments));
+  return true;
+}
+
+auto RewriteBuilder::VisitUnaryOperator(clang::UnaryOperator* expr) -> bool {
+  switch (expr->getOpcode()) {
+    case clang::UO_AddrOf:
+      Write(expr, {OutputSegment::Text("&"),
+                   OutputSegment::Node(expr->getSubExpr())});
+      break;
+
+    default:
+      // TODO: Finish implementing cases.
+      break;
+  }
+  return true;
+}
+
+auto RewriteBuilder::VisitVarDecl(clang::VarDecl* decl) -> bool {
+  // TODO: Check storage class. Determine what happens for static local
+  // variables.
+
+  bool is_const = decl->getType().isConstQualified();
+  std::vector<OutputSegment> segments = {
+      OutputSegment::Text((llvm::Twine(is_const ? "let " : "var ") +
+                           decl->getNameAsString() + ": ")
+                              .str()),
```

Comment:
```suggestion
      OutputSegment::Text(llvm::formatv("{0} {1}: ", is_const ? "let" : "var", decl->getNameAsString()),
```

Just suggesting a different approach which might be easier to read at a glance.

Reply:
Done.